### PR TITLE
type check in InferenceModel.get_embeddings()

### PR DESCRIPTION
**Pull request by yutanakamura-tky**
### Overview
I added type check so that `InferenceModel.get_nearest_neighbors()` can accept a list of strings (=a batch of text data) as a variable `query`.

### Detail
`InferenceModel.get_nearest_neighbors()` calls `get_embeddings()`.

When a batch is given as a list of tensors, `get_embeddings()` stacks them. However, this behavior causes an error when a batch is given as a list of string (=a batch of text data):

```python
    def get_embeddings(self, x):
        if isinstance(x, list):
            # any list will be passed here
            x = torch.stack(x)
        ...
```

More fine-grained type check can avoid this error by skipping `torch.stack()` and going directly to the following procedures.

```python
    def get_embeddings(self, x):
        if isinstance(x, list):
            if all(map(lambda x_: isinstance(x_, torch.Tensor), x)):
                # Only a list of torch.Tensor will undergo this
                x = torch.stack(x)
        ...
```

### Example

This Pull Request will help to evaluate the performance of metric learning to find images corresponding to a query text.

```python
class MyImageTextEncoder(torch.nn.Module):
    def __init__(
        self,
        image_encoder: torch.nn.Module,
        text_encoder: torch.nn.Module,
        mode:Literal['pair', 'image_only', 'text_only']
    ):
        self.image_encoder = image_encoder
        self.text_encoder = text_encoder
        self.mode = mode

    def forward(
        self,
        batch:Union[
            Tuple[torch.Tensor, List[str]],
            torch.Tensor,
            List[str]
        ]
    ) -> Union[Tuple[torch.Tensor, torch.Tensor], torch.Tensor]:
        if self.mode == 'pair':
            images, texts = batch
            return (self.image_encoder(images), self.text_encoder(texts))
        elif self.mode == 'image_only':
            images = batch
            return self.image_encoder(images)
        elif self.mode == 'text_only':
            texts = batch
            return self.text_encoder(texts)

image_dataset = MyImageDataset()  # each sample is Tuple[torch.Tensor, int]
text_dataset = MyTextDataset()  # each sample is str
text_dataloader = torch.utils.data.DataLoader(text_dataset)

image_text_encoder = ImageTextEncoder()
match_finder = MatchFinder()

# Switched mode so that image_text_encoder() takes images: torch.Tensor and returns torch.Tensor
image_text_encoder.mode = 'image_only'
inference_model = InferenceModel(
    trunk = image_text_encoder,
    match_finder = match_finder
)
inference_model.train_indexer(image_dataset)

# Switched mode so that image_text_encoder() takes texts: List[str] and returns torch.Tensor
image_text_encoder.mode = 'text_only'
results = []
for batch in text_dataloader:
    # batch is List[str]
    result = inference_model.get_nearest_neighbors(batch, k=10)
    results.append(result)
```

### CODE REVIEW

File: src/pytorch_metric_learning/utils/inference.py [hunk 1]
```diff
@@ -93,11 +93,9 @@ def __init__(
         self.normalize_embeddings = normalize_embeddings
 
     def get_embeddings_from_tensor_or_dataset(self, inputs, batch_size):
-        if isinstance(inputs, list):
-            inputs = torch.stack(inputs)
-
+        inputs = self.process_if_list(inputs)
         embeddings = []
-        if torch.is_tensor(inputs):
+        if isinstance(inputs, (torch.Tensor, list)):
```
Comment (KevinMusgrave) [inline src/pytorch_metric_learning/utils/inference.py:98]: Treat input as a batch of batches if input is a tensor **or list**, i.e. if the input has length 128, and the batch_size is 32, then the model will consume 4 batches.

**Comment by KevinMusgrave** [comment]
Looks good. I made some minor changes. The functionality is the same, except for line 98. See the comment above
